src/wp-includes/sitemaps/class-wp-sitemaps-registry.php

```diff
@@ -0,0 +1,91 @@
+<?php
+/**
+ * Sitemaps: WP_Sitemaps_Registry class
+ *
+ * Handles registering sitemaps.
+ *
+ * @package WordPress
+ * @subpackage Sitemaps
+ * @since 5.5.0
+ */
+
+/**
+ * Class WP_Sitemaps_Registry.
+ *
+ * @since 5.5.0
+ */
+class WP_Sitemaps_Registry {
+	/**
+	 * Registered sitemaps.
+	 *
+	 * @since 5.5.0
+	 *
+	 * @var array Array of registered sitemaps.
+	 */
+	private $sitemaps = array();
+
+	/**
+	 * Maximum number of sitemaps to include in an index.
+	 *
+	 * @sincee 5.5.0
+	 *
+	 * @var int Maximum number of sitemaps.
+	 */
+	private $max_sitemaps = 50000;
+
+	/**
+	 * Adds a sitemap with route to the registry.
+	 *
+	 * @since 5.5.0
+	 *
+	 * @param string               $name     Name of the sitemap.
+	 * @param WP_Sitemaps_Provider $provider Instance of a WP_Sitemaps_Provider.
+	 * @return bool True if the sitemap was added, false if it is already registered.
+	 */
+	public function add_sitemap( $name, $provider ) {
+		if ( isset( $this->sitemaps[ $name ] ) ) {
+			return false;
+		}
+
+		if ( ! $provider instanceof WP_Sitemaps_Provider ) {
```

Comment:
Another great side effect of using typehints above is that we'll be able to remove old-school clauses like this.